### ISSUE TITLE
single-file workflow - behavior without a configuration file

### ISSUE THREAD

**Issue by MyriaCore** [entangled/entangled#66]
This is going to be a weird proposal, so just bear with me here.

### TL;DR

Implementing the following features would allow entangled to be more effective when used on "projectless" markdown files:

1. add reasonable defaults so entangled can run without the dhall file
2. add command-line switches that allow the user to reassign `entangled.Config` per-invocation
3. parse `entangled.Config` fields in the front matters of markdown documents

### Use Case

Quite often, programmers will doodle around in a single source file. This can actually be an efficient way to explore some ideas, given that the tooling doesn't *require* project-level infrastructure, such as configuration files in the working directory. This is doubly so for languages with strong repl workflows, that can quickly load single-files into a repl and allow you to play around with the local definitions. A programmer could just throw a source file somewhere, add some logic to it, and debug the logic by loading it into the language's repl, all without having to setup a full project with full tooling. 

Some remarks about this usecase:
- While you might not think daemonized tooling works in this usecase, it *does*, the only restriction is dependency on local files or environment variables in the environment.
- This usecase in fact could see quite a bit of benefit, since you could theoretically generate an entire project from a single literate source file. 

### The Problem

Currently, entangled does **not** support this use-case, as it requires an (albeit minimal) configuration file to operate. The [example configuration](https://github.com/entangled/entangled/blob/df0e5c0c46ec0e80dc842f827b6eb3a4a518debd/data/example-config.dhall) appears geared towards a more project-oriented mindset, so it wouldn't work for this use-case. While the [minimal configuration](https://github.com/entangled/entangled/blob/df0e5c0c46ec0e80dc842f827b6eb3a4a518debd/data/minimal-config.dhall) itself is good for the purposes of the single-file development workflow, it *still* breaks this workflow, because it's still a configuration file sitting in the working directory.

To support this workflow, a solution that eliminates the need for a configuration file is a must-have. For this, I have multiple proposals, which we *could* pick and choose from, but are kinda designed to work well with each other in a single implementation:

### Proposal 1: Assume 'reasonable' defaults when using without a command-line switch

Currently, entangled refuses to run with most of its subcommands whenever there's no `entangled.dhall` file in the working directory, but this need not be the case. As we talked about, the [minimal configuration](https://github.com/entangled/entangled/blob/df0e5c0c46ec0e80dc842f827b6eb3a4a518debd/data/minimal-config.dhall) would work as a good base, the only thing we'd need to do here is have some reasonable defaults.

In this case, I think it's reasonable to suggest that the default `watchList` should be `["*.md"]`, which would target all markdown files in the working directory. If some of these files don't contain meaningful chunks, they won't produce any code files, so you don't get any junk in the filesystem. It may be tempting to borrow the `["lit/*.md"]` from the [example configuration](https://github.com/entangled/entangled/blob/df0e5c0c46ec0e80dc842f827b6eb3a4a518debd/data/example-config.dhall), but a user is less likely to be using a single-file workflow if they've already got a `lit` folder with markdown documents in it.

Alternatively, I think pre-existing `.entangled/db.sqlite` is just file for the `database` entry. 

This setup would at least allow users to start tangling and stitching their literate markdown files, which is what we want.

### Proposal 2: Override configuration variables via command-line switches

This might be a nice feature to have in general. 

It might work similar to java's `-D<jvm var>` switches. We could use these to allow users to re-assign record variables via the command line. For example, if `-C` was our switch prefix, we could set the database for a given command with the following invocation:

```shell
$ entangled -Cdatabase 'Some ".entangled/db.sqlite"' <command>
```

You could imagine invocations that set children record variables as well:

```shell
$ entangled -Csyntax.matchCodeStart '"```[ ]*{[^{}]*}"' <command>
```

Because this is reassignment, there's nothing stopping users from running entangled commands that modify variables in the existing configuration:

```shell
$ entangled -Clanguages '{ name = "Unlambda", identifiers = ["unlambda"]
                         , comment = entangled.comments.hash
                         } :: entangled.languages' \
            <command>
```

At the end of the day, the expressions given to these `-C` switches should just be dhall expressions. 

### Proposal 3: Configuration via Front Matter

Configuration could be expressed in the frontmatter of a document. The most common type is YAML front matter, so you could imagine a minimal configuration looking something like:

```yaml
title: 'Very Important Document'
author: 'Joe Jackson'
entangled:
  database: '.very-important-db.sqlite'
```

One thing to note is that if a `watchList` entry is not provided, we can still reasonably assume that the `watchList` **at least** contains the `Very Important Document`.  

There are a few strategies we could employ to recognize *which* document contains the configuration in the absence if an `entangled.dhall` file:

1. Loop through the markdown documents in the working directory, and pick the first one with a schema-compliant `entangled` entry in its frontmatter
2. Add a command-line switch that could specify the configuration file to use, such as `-c` or `--config-file` (maybe `-c` for `--check` could be turned to `-ch` instead?). This would honestly be nice to have *aside* from the single-source use case. Semi-related to proposal 2.
3. Check files with 'important names' first. File names like `README.md` and `readme.md` could be checked first because in a project setup, the user would likely have their relevant settings in a root-level file like the readme. This could both enable easy production of example snippets in the readme, *and* it could allow for very minimal tarballs to be distributed, under the assumption that building the repository will be done by running entangled first to generate the source.

My only remark here is that we should be wary when implementing this, so we can still have a way to provide document-specific overrides of settings in the future.

**Comment by jhidding** [comment]
first quick reaction: wow, I love it; I will break this down into a project.

## Proposal 1: default config
- I've been meaning to have an in-memory default database at some point.
- The default `watchList` should be empty. If you use this for experimenting, like you propose, you may have a folder full of loose markdown files that you don't want to mix. I would prefer, for this case a cmd-line argument (shorter than `-CwatchList "[experiment42.md]"`, probably just `-i experiment42.md`).

## Proposal 2: command-line config
- This looks nice, but I'm not yet sure what is the best way to implement this with respect to Dhall. This assumes an implicit `let entangled = <current default schema>` before any `-C` statement. With multiple statements we need to combine all of these into the existing config. We need to formalise how this is done by specifying the 'effective' Dhall config file that you'd obtain.

## Proposal 3: front matter config
This is the most complex of your proposals. However, I like the way this would make a code completely self-contained. This makes it worth the extra effort.
- Dhall expands to JSON, so we could read the front matter `entangled` item and use the existing Dhall schema for this. We should formalize how config is added to existing config: some lists should be appended, other items should be overrides.

1. If I'm being liberal:
    - I'm wary of making too many assumptions on how this is used. If we consider the collection of all markdown files in the watchList, and a way of summing the configurations, we could see if there are no conflicts in the header sections. If there are conflicting settings, throw an error and exit the daemon. I can imagine having one chapter where you'd like to introduce a new language (that entangled doesn't know about by default), then you might want to add that specific config information to that document, and not the main one.
    - Since documents can cross-reference code between them, settings should be global. One exception is the `syntax` setting. I could envision an override for that being useful when parsing the markdown; after which the override is discarded.

2. If I'm being pedantic:
    - we have two use-cases, single-file and project. In the case of a project, I think the 'one-entangled.dhall-to-rule-them-all' approach fits best to current best practices: all the major tools use this approach. This leaves no room for confusing behaviour, config in many places influencing behaviour globally etc.
    - In the case of single-file, it is nice to be able to have settings in the header, and we don't need to worry about other settings floating around elsewhere.

I'm leaning towards the pedantic here.

**Comment by MyriaCore** [comment]
Iterating on a few things you mentioned:
1. An in-memory default database would certainly make this workflow easier. I think that having a `-i` does work, in proposal 3 I suggested a `--config-file` or `-c` command, but if we're being pedantic (as you mentioned in your analysis of proposal 3), then we can probably safely confine ourselves to a `-i` when in single-file mode. 
2. I agree, this might be the weakest of the three proposals. I was thinking you'd use this in CI/CD for example, where you could pass minimal configurations to as command line arguments so you could see what's happening when your document is being tangled. However, at that point you could easily have an `entangled.dhall` file, or fall back on the first or third proposal here.
3. I'm kinda leaning towards the pedantic implementation too, tbh. As I said before, we could indicate running in single-file mode with the `-i` you suggested, since that's the only time you'd really use it. 

My final suggestion is that we should consider replacing `-i` (I assume this is shorthand for `--input`) with something that would suggest you're running in single-file mode. Maybe `-i` is for `--single-input`? I guess that would be fine, but I feel like `-s` might work well too there, since it could double as *'script'* (which is probably the most common type of single-input file), or *'single-input'*.
